### PR TITLE
[READY] Ammunition management for ballistic mech weapons, and a nerf to mech missiles

### PR DESCRIPTION
**Pull request by zxaber**
## About The Pull Request
Adds an ammo system and printable ammunition to most mech weapons with a "rearm" function. Replaces the old system of generating ammo at a power cost. With the exception of the missile launcher and clusterbang launcher, all affected weapons have a storage capacity of ammunition equal to four-times the amount of ammo that the weapon can load (though the storage cache starts with only one full reload when printed from the mechfab). Ammo boxes are printed from the mechfab and also are available in the Sec protolathe, and each box will carry exactly one full reload. Ammo boxes are not terribly expensive, but reloading requires either an ally on foot to pop the ammo box in, or for you to hop out and do it yourself.

Missile tubes and the clusterbang launcher do not have an ammo storage cache, and so reloading them will reload the tubes directly. This is not the full extent of nerfs that the missiles need ~~(I will do those in a separate PR since I suspect this one will be controversial)~~ (read below), but removing the ability to reload missile pods without an ally or exiting the mech may help a bit.

The SRM-8 has been replaced with the new BRM-6 in tech webs. The BRM carries a smaller explosion and does not explode unless hitting a wall, door, shutter, or another mech, and only has six shots. The SRM-8 remains on the Mauler mech. If the BRM missile does not explode, it will deal 30 brute damage instead, and break.

## Why It's Good For The Game
Currently, mech ballistic weapons function like reloadable lasers. You have an ammo counter, but it's one hundred percent based on your energy cell. This gives some flavor to ballistic weapons, making them actually feel different, and gives them a mini-niche in the form of good choices to use alongside an energy-draining mech ability (like the Gygax speed).

Use lasers when you won't need to worry about power and/or need burn damage.
Use ballistics when you want to save power and/or need brute damage.

The SRM-8 to BRM-6 change is being done here so that I don't have to wait until this PR is merged before starting it. The idea is to keep the reasons why you'd want to build a missile launcher (wall breaking and mech fighting) without also outclassing most other weapons burst damage potential. I'm not truly happy that the explosions deal zero damage to surrounding people even one tile away, but dealing no damage is better than the massive damage + stun that the current missile launcher does. Nuke Mauler has the old launcher, since I believe the consensus is that Nukie Mechs are already not generally worth the money.

## Changelog
:cl:
balance: Mecha ballistics weapons now require ammo created from an Exosuit Fabricator or the Security Protolathe, though they will start with a full magazine and in most cases enough for one full reload. Reloading these weapons no longer chunks your power cell. Clown (and mime) mecha equipment have not changed.
balance: The SRM-8 Missile Launcher has been replaced with the BRM-6 Breaching Missile Launcher in techwebs (Nukie Mauler remains equipped with the SRM-8).
balance: Both Missile Launchers and the Clusterbang Launcher do not have an ammo cache, and cannot be reloaded by the pilot. Once the initial loaded ammo has been spent, you can use the appropriate ammo box to load the weapon directly.
add: Utility mechs that have a clamp equipped can load ammo from their own cargo hold into other mechs.
add: Nuke Ops can purchase spare ammo duffel bags for their mecha weapons, should they run low.
/:cl:

To-Do:
- [x] Add ammo to security protolathe
- [x] Get better ammo box sprites (hopefully)
- [x] Possibly adjust ammo counts for balance. Might make the flashbang launcher and/or cluster launcher behave like the missile pod.
- [x] Add functionality for the utility mechs to use clamps to load ammo from their cargo hold into another mech.

Ammo counts, for anyone wondering:

Weapon | Magazine Size | Total ammo at start* | Maximum total ammo*
------------ | ------------ | -------------| -------------
Scattershot† | 40 (10) | 80 (20) | 200 (50)
Hades | 24 | 48 | 120
Ultra AC 2† | 300 (100) | 600 (200) | 1500 (500)
BRM-6 Missile Rack | 6 | 6 | 6
SRM-8 Missile Rack | 8 | 8 | 8
Flashbang Launcher | 6 | 12 | 30
Cluster Launcher | 3 | 3 | 3

*Total ammo is the amount in the weapon's magazine plus the stored ammo.
†The Scattershot lists its ammo as 40, but uses four per shot. So in reality, a single full reload gives you ten shots. Similarly, the UAC2 fires in three-round bursts, and so 300 rounds is 100 bursts.

Clown weapons (and the mime weapon) are all unaffected by this change.

How it's displayed to the pilot:
image

### CODE REVIEW

**Comment by Dawson1917** [comment]
Very needed nerf that might do something to stem SRM8 - AC2/Solaris - cluster/utility meta build for mechs and missile spam. As it is the energy required for reloading weapons is laughable and never a problem since most mechs will use bluespace or yellow cells.

**Comment by necromanceranne** [comment]
Neat! Are you going to implement that ripley reloading thing mentioned in the other thread? That sounded cool.

Also, the real balance point of ballistics verses lasers is window shooting. Infrequently exploited, windows/see through doors make for fantastic shields against ions and ballistic weapons. The Solaris cannon is fantastic for this reason despite the energy usage. But the problem was mostly that ballistics just did more damage overall. And that's why they were used more because they came with no real downsides and reloading was one button press every so often.

A positive change. Makes the weapon choices actually matter.

**Comment by zxaber** [comment]
I do want to add loader mechs being able to restock combat mechs. I'm undecided on if I want to tack on the code to the clamp, or if it's worth having a dedicated equipment (even if I like the clamp idea better). It may or may not be in this PR, but I hope to have mech-to-mech restocking done soonish in some fashion.

**Comment by Timonkeyn** [comment]
maybe make it a mod of the mech to generate ammo at Energy Cost like the reflective armor for the mech Thing? would that be balanced?

**Comment by Arathianen** [comment]
Great and intuitive change!

I am just a bit confused about the missile. Is it 8+ 8 stored or will the total amount be 8?

**Comment by zxaber** [comment]
@Timonkeyn 
Having an equipment to basically turn ballistics back into glorified lasers is what I'd like to avoid. 

I could see perhaps a whole new combat mech who's _special ability_ is to generate ballistics ammo, but that's a topic for another PR. The point of this one is to make ballistics not act like lasers.

@Arathianen 
It means 8 total. The missile launcher has a storage capacity of zero missiles, so it's 8 loaded + 0 max storage.

The take-away is that when you load any other weapon, you're putting ammo into the storage bin for the auto-loader to feed into the weapon. The missile launcher is just eight tubes, so when you hit it with the ammo box, it loads the tubes directly.

**Comment by zxaber** [comment]
So I had some inspiration on how the clamp loading should work, and I managed to get that squared away. APLU mechs with a clamp can now support combat mechs by loading more ammo in. Each use will load a single weapon with up to a single box's worth of ammo, so you may need multiple clicks. You'll get a message when you can't load any more on a particular mech (along with the normal message and audio you'd also get on foot if it loads successfully).

Additionally, got security's protolathe sorted out. Ammo for mech ballistic weapons will start showing up once the tech is researched. Each ammo box is on the same node as the weapon it's designed for.

**Comment by HomocidlePancakes** [comment]
What about nukie mechs? Will they have higher ammo capacity, or maybe have you able to purchase ammo?

**Comment by zxaber** [comment]
Hmm, that is a good point. I'll have to look into it; maybe I can at least fill their ammo caches up.

**Comment by ATH1909** [comment]
> I could see perhaps a whole new combat mech who's _special ability_ is to generate ballistics ammo, but that's a topic for another PR.

That'd be hella cool, and I'd totally be down with that mech idea (maybe make it require an anomaly core or two, though?).

**Comment by zxaber** [comment]
>maybe make it require an anomaly core or two, though?

Maybe, but there *would* already be the opportunity cost of not getting to use Gygax speed, Phazon noclip, or (if it ever gets buffed) Durand defense. Either way, a thought for another PR.

**Comment by zxaber** [comment]
@HomocidlePancakes 
Nuke ops mechs now start with max ammo, which is enough for four full reloads plus the ammo pre-loaded into the actual weapon. If need be, I can easily just increase the size of ammo storage on weapons that get auto-attached by nuke mech Initialized() procs.

The exception for this is the missile launcher on the Mauler. The missile launcher does not have backup ammo, and is intended to only get the eight missiles before needing more ammo printed. If this is an issue for ops, I can make a nuke-ops variant that uses ammo storage like other weapons. Truth be told, I'm planning a re-work to how the missile works (possibly pushing away unanchored mobs and items with medium damage while hitting turf and anchored objects like mechs hard). So we may end up wanting to give the Mauler a different weapon in this slot anyway.

**Comment by necromanceranne** [comment]
Maybe add ammo into the nukie uplink? Just to be on the safe side? It's like a reload for the l6. You are probably not going to need it but it is better to have it than find yourself running low just because you had to cut your way through the vast majority of the crew yourself.

**Comment by HomocidlePancakes** [comment]
I haven’t read the code, but I will assume that every weapon has it’s own refill packs. How about adding a specifc “ammo” category  to the exofabs? Will make it easier to navigate

**Comment by zxaber** [comment]
Already set. I have the Exosuit Ammunition category sitting between Exosuit Equipment and Cyborg Upgrades. It'll populate at the same time the weapons get researched.

**Comment by GuyonBroadway** [comment]
Add ammo explosions that fuck up the pilot battletech style!

**Comment by zxaber** [comment]
We only _just_ got rid of explosions when the mech is killed, and random malfunctions from taking damage is already terrible. I'd rather not have an extra downside to ballistics on top of the fact that they're more work to upkeep than lasers are.

@necromanceranne
How should I price them? A TC for just eight rockets seems expensive, but because the boxes are bulky, giving them several boxes would be inconvenient to carry. Should I go with a "Dark gygax spare ammo" duffel bag with like three boxes of each type, and have likewise for the Mauler?

**Comment by GuyonBroadway** [comment]
As an aside logical chain thinking here.

Missles are bad.
They are bad because they stun you instantly for a long time even if they miss slightly.
The stun comes as a result of missiles causing an explosion which stuns if it hits you.
All explosions cause this KO effect.

Suggestion, gut the fuck out of rockets or maybe even the stuns that come with explosions.

**Comment by zxaber** [comment]
Missiles will have a proper nerf PR soon (tm)

**Comment by bigfatbananacyclops** [comment]
reeeeeee noooooo

**Comment by CPTANT** [comment]
I would love to see the missile ammo count reduced to 4. Those missiles are way to spammable for how destructive they are.

**Comment by zxaber** [comment]
Alrighty, changes:
- SRM-8 Missile Launcher has been replaced with the new BRM-6 Missile Launcher in techwebs. The BRM is a breach missile launcher, and missiles will only explode if they hit a wall or other mech (or in the words of the weapon's description, "designed to explode only when striking a sturdy target"). The explosion is also much smaller, and does not damage anyone standing to the side. The result is a missile that is great for taking down walls quick, or dealing damage to mechs, but do little else. Shooting anything non "sturdy" results in 30 brute damage and a broken missile dropping to the ground (the latter is for fluff, really). As the name implies. the BRM-6 also only has a capacity of six shots, and the sprite for the launcher has been edited to reflect that. The SRM-8 is still in the game, and is still attached to the Nuke Ops Mauler, but is otherwise unobtainable for station crew.
Ammo boxes for the two missile types are not interchangeable, but have slightly different sprites.

- Added in ammo bundles for Nuke Ops mechs. These come in duffel bags, and each bag includes three full reloads for each of the mech's weapons. The Dark Gygax bag, which services two weapons, is 4 TC, while the Mauler for three weapons is 6 TC. This is a bit low compared to other ammo costs, but I didn't want to increase the cost of fielding a Nuke mech, as they already seem somewhat underpowered.

- Clusterbang launcher has been changed to not have an ammo cache. Like the missile launchers, it will have a max ammo of three shots before needing to be re-armed via an ammo box. I don't see these used much, but truth be told they seem super terrible to fight against unless you have both eye and ear protection. It's like eighteen flashbangs dropped after a short delay. They'll still be great for clearing rooms, but you won't get to spam them.

I wanted to do the missile launcher change in its own PR, but since it includes changes in this PR (like the ammo variables), I would have had to wait until this one got merged. Hopefully no one minds me tacking it on here.

**Comment by zxaber** [comment]
New ammo box sprites;

image

Top is an empty box. Clockwise from there is ammo for the LMG, BRM-6, Flashbang Launcher, Scattershot, FNX-99 "Hades", SRM-8, and Clusterbang Launcher.

Once I commit them, and do some last-minute testing, I think I can set this PR to ready.

File: code/game/mecha/equipment/weapons/mecha_ammo.dm [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+	to_chat(user, "<span class='notice'>You fold [src] flat.</span>")
+	var/I = new /obj/item/stack/sheet/metal(user.loc)
+	qdel(src)
+	user.put_in_hands(I)
+
+/obj/item/mecha_ammo/examine(mob/user)
+	..()
+	if(rounds)
+		to_chat(user, "There [rounds > 1?"are":"is"] [rounds] [round_term][rounds > 1?"s":""] left.")
```
Comment (optimumtact) [inline code/game/mecha/equipment/weapons/mecha_ammo.dm]: examine is now a list
```
. = ..()
. += "There [rounds > 1?"are":"is"] [rounds] [round_term][rounds > 1?"s":""] left."
```

**Comment by optimumtact** [comment]
Minor concern but otherwise this LGTM

File: code/game/mecha/equipment/weapons/mecha_ammo.dm (hunk 1 above)
Comment (zxaber) [reply code/game/mecha/equipment/weapons/mecha_ammo.dm]: Doesn't work on my build, since it's behind, but I have it set now.

**Comment by FloranOtten** [comment]
So... why do we need this? Mechs already have a massive nerf. You can now counter mechs by literally just spamming ions. Why do we need another massive nerf? Before you go 'this isn't a big nerf!' yes, it is. No-one uses anything but ballistics because the laser guns are bad. The only proper weapons that are non ballistic mechs have are:

Ion cannon (for borgs and mechs)
Tesla gun

**Comment by GuyonBroadway** [comment]
Tg does balancing like blizzard entertainment does balancing.

Things are either overpowered and "the meta" or weak to the point of uselessness.

**Comment by Dawson1917** [comment]
I like the missile nerf, but what's the point of breaking walls with a limited ammunition missile launcher when the gygax and durand can already punch down even r-walls easily? A solid anti-mech weapon is good but how often do you see mechs fight against eachother? Either the roboticist builds a mech to go on a killing spree and destroys the two fabricators so no other mechs can be built, or they come back to robotics for recharging/etc in the long time it takes to build a mech and kill the person(s) building another mech. I see this being most useful during warops on the off-chance nukies order a mech, and robotics (and rnd, and mining) are hurrying and manage to make a gygax before the 20-minute mark. This is personally something I've never seen. 
To me a missile launcher should be useful as an AOE damage weapon without the cancer of 3x3 sleep stuns or instant critting, rather than only taking out walls and other mechs

**Comment by zxaber** [comment]
Far as I can tell, the effect of knocking people down for ages is just how explosions work. I can call the explosion call with different sizes, but even the lightest damage radius knocks you over. So it's either make it a breach weapon or just remove it altogether.

**Comment by Dawson1917** [comment]
if you ever want to add another missile weapon to the game that isn't awful to play against like the SRM-8 consider a shrapnel missile. damage in a 3x3 area from the shrapnel, lands where you hit, doesn't do explosion damage. that gives it utility as a weapon against crowds of people, something lacking in many mech weapons, and the damage (20 damage in each tile? 30 damage? depends how slow you want the fire rate) is spread over the 9 tiles so you can't 1-hit a single target but damage multiple people in one go instead. explosive missiles are hopeless unless sleep stuns from explosions are nerfed but that would also nerf welderbombs and syndicate minibombs quite hard. up to you

**Comment by zxaber** [comment]
So basically a scattershot with a ranged split point. That'd be neat. I might see about creating that in a future PR.

I suppose if the explosive proc got an extra input that allowed callers to specify no sleep or knock over, then we could add a small bit of damage radius to the breaching missile. But a change to explosion() is outside the scope for this PR.

**Comment by zxaber** [comment]
@FloranOtten 
>So... why do we need this? Mechs already have a massive nerf. You can now counter mechs by literally just spamming ions. Why do we need another massive nerf? Before you go 'this isn't a big nerf!' yes, it is. No-one uses anything but ballistics because the laser guns are bad. The only proper weapons that are non ballistic mechs have are:

>Ion cannon (for borgs and mechs)
>Tesla gun

It adds some extra work from time to time to get more ammo, but you don't lose a chunk of power every time you reload. It's a flat nerf only if you were using a yellow cell, missiles, or clusterbangs.

Anyway, the ammo boxes are _really_ not that expensive, and you'll just need to return to robotics (or security) occasionally to resupply. You can carry a total of five mags-worth of ammo, including what is loaded in the weapon itself, so you won't need to return after every single fight unless you're banking on missiles or clusterbangs. And if you do find yourself running low constantly, consider making a Ripley with a posi/AI pilot and a clamp; APLU mechs can load ammo into other mechs from their cargo hold.

Honestly, ballistics weapons being glorified lasers before was just silly. If you want ballistics weapons, you deal with ammo logistics; not having to keep track of ammo is the _whole point_ of lasers in the first place.

**Comment by zxaber** [comment]
One last change (hopefully); breaching rockets now use a list of possible "sturdy" types, and doors and shutters are now on the list of valid targets.

**Comment by optimumtact** [comment]
This is a nerf yes, but it actually increases the detail of the simulation and that's cool

**Comment by GuyonBroadway** [comment]
Suggestion, battletech style heat mechanics.

Because having a laser cannon with the fire rate of a laser gun is funny until the control consoles melt.

**Comment by zxaber** [comment]
Ideally, the energy usage does the purpose of heat. If lasers end up dominating because of this change, I might consider adding heat mechanics _alongside some further balancing_.

After this PR, I'd really like to let mechs sit for a bit to allow people time to adjust and really see where the balance lies. I may look into doing a rework on the Durand's ability, but that's it.

**Comment by FloranOtten** [comment]
>It adds some extra work from time to time to get more ammo, but you don't lose a chunk of power >every time you reload. It's a flat nerf only if you were using a yellow cell, missiles, or clusterbangs.

Have you considered a career in politics? That's some grade A bullshit. The chunk of power was such a minor thing that nobody even noticed it except during the most drawn out engagement. Sure, it's technically not a flat nerf, the same way that removing the Xenobiology console but also making slimes *slightly* less hostile is technically not a flat nerf.

**Comment by optimumtact** [comment]
Stop shitslinging floran or I will sling you right out of this codebase

File: code/game/mecha/mecha.dm [hunk 2]
```diff
@@ -1106,3 +1106,53 @@ GLOBAL_VAR_INIT(year_integer, text2num(year)) // = 2013???
...
+
+///////////////////////
+////// Ammo stuff /////
+///////////////////////
+
+/obj/mecha/proc/ammo_resupply(var/obj/item/mecha_ammo/A, mob/user,var/fail_chat_override = FALSE)
+	if(!A.rounds)
+		if(!fail_chat_override)
+			to_chat(user, "<span class='warning'>This box of ammo is empty!</span>")
+		return 0
```
Comment (optimumtact) [inline code/game/mecha/mecha.dm]: TRUE/FALSE defines please

Comment (optimumtact) [reply code/game/mecha/mecha.dm]: there's a few places they could apply

Comment (zxaber) [reply code/game/mecha/mecha.dm]: Done and done

**Comment by zxaber** [comment]
oof

**Pull request merged**